vllm/v1/worker/gpu_model_runner.py

```diff
@@ -799,6 +799,21 @@ def execute_model(
                     # This relies on cuda-specific torch-internal impl details
                     generator.set_offset(generator.get_offset() - 4)
 
+        # num_reqs entries should be non-None
+        assert all(
+            req_id is not None for req_id in
+            self.input_batch.req_ids[:num_reqs]), "req_ids contains None"
+        req_ids = cast(List[str], self.input_batch.req_ids[:num_reqs])
+
+        # NOTE: GPU -> CPU Sync happens here.
```

Comment:
Just for a record: If top-p or top-k sampling is used (with the FlashInfer kernel), CPU-GPU synchronization happens inside the sampler at https://github.com/vllm-project/vllm/blob/324960a95c00112ce6b9b858d9311da1597cfb8b/vllm/v1/sample/ops/topk_topp_sampler.py#L193-L194

Reply:
Do you think we can avoid this in a follow up PR?

Reply:
I don't think we can. This is a fundamental limitation of the kernel (or the algorithm itself). The rejection sampling method cannot 100% guarantee the success.